src/OpenBinary/reader.jl

```diff
@@ -453,29 +448,123 @@ function PSRI.goto(graf::Reader, t::Integer, s::Integer = 1, b::Integer = 1)
     return nothing
 end
 
-function _get_position(graf, t::Integer, s::Integer, b::Integer)
-    if PSRI.is_hourly(graf)
+function _get_relative_offset(ptr::Reader)
+    return 4 * ptr.agents_total * ptr.scenario_total * (ptr.first_relative_stage - 1)
+end
+
+function _get_relative_offset(::Any)
+    return 0
+end
+
+function _get_position(ptr::Any, t::Integer, s::Integer, b::Integer)
+    relative_offset = _get_relative_offset(ptr)
+
+    if PSRI.is_hourly(ptr)
         # hours in weekly = 52 * 168 = 8736
         # hours in monthly = 8760
         pos = 4 * (
-            graf.blocks_until_stage[t] * graf.agents_total * graf.scenario_total +
-            (s - 1) * graf.agents_total * graf.blocks_per_stage[t] +
-            (b - 1) * graf.agents_total)
-        return pos + graf.hs
+            ptr.blocks_until_stage[t] * ptr.agents_total * ptr.scenario_total +
+            (s - 1) * ptr.agents_total * ptr.blocks_per_stage[t] +
+            (b - 1) * ptr.agents_total)
+        return pos + ptr.offset - relative_offset
     else
         pos = 4 * (
-            (t - 1) * graf.agents_total * graf.block_total * graf.scenario_total +
-            (s - 1) * graf.agents_total * graf.block_total +
-            (b - 1) * graf.agents_total)
-        return pos + graf.hs
+            (t - 1) * ptr.agents_total * ptr.block_total * ptr.scenario_total +
+            (s - 1) * ptr.agents_total * ptr.block_total +
+            (b - 1) * ptr.agents_total)
+        return pos + ptr.offset - relative_offset
+    end
+end
+
+function _get_first_stage_date(stage_type::PSRI.StageType, _year::Integer, _first_stage::Integer)
+    if stage_type == PSRI.STAGE_MONTH
+        @assert 1 <= _first_stage <= 12
+        
+        return Dates.Date(_year, 1, 1) + Dates.Month(_first_stage-1)
+    elseif stage_type == PSRI.STAGE_WEEK
+        @assert 1 <= _first_stage <= 52
+
+        return Dates.Date(_year, 1, 1) + Dates.Week(_first_stage-1)
+    elseif stage_type == PSRI.STAGE_DAY
+        @assert 1 <= _first_stage <= 365
+        ret = Dates.Date(_year, 1, 1) + Dates.Day(_first_stage-1)
+
+        if Dates.isleapyear(_year) && _first_stage > 24 * (31 + 28)
+            ret += Dates.Day(1)
+        end
+        
+        return ret
+    else
+        error("Stage type '$(stage_type)' is not currently supported")
     end
 end
 
+function _get_hour_discretization(stage_type::PSRI.StageType, block_total::Integer)
+    hour_total = if stage_type == PSRI.STAGE_WEEK
+        168
+    elseif stage_type == PSRI.STAGE_MONTH
+        if block_total % 672 == 0
+            672
+        elseif block_total % 720 == 0
+            720
+        else
+            744
+        end
+    elseif stage_type == PSRI.STAGE_DAY
+        24
+    elseif stage_type == PSRI.STAGE_YEAR
+        8760
+    else
+        error("Stage type $(stage_type) is not currently supported")
+    end
+
+    return floor(Int, block_total / hour_total)
```

Comment:
Not that it makes a huge difference but... https://docs.julialang.org/en/v1/manual/performance-tips/#Tweaks 
try to use `fld`

Reply:
@joaquimg @guilhermebodin Taking a closer look at this function I ask myself: Why keeping `ptr.is_open` instead of calling `Base.isopen` on `ptr.io`?

Reply:
This function is called once per load. Conceptually, `div` is better suited since operands are integers.